File: algorithms/257. Binary Tree Paths/solution.cc
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<string> binaryTreePaths(TreeNode* root) {
        vector<string> paths;
        dfs(root, "", paths);
        return paths;
    }

    void dfs(TreeNode * root, string p, vector<string> & paths) {
        if ( NULL == root ) {
            return;
        }
        if ( NULL == root->left && NULL == root->right ) {
            paths.push_back(p + to_string(root->val));
        }
        p = p + to_string(root->val) + "->";
        dfs(root->left, p, paths);
        dfs(root->right, p, paths);
    }
};

File: algorithms/257. Binary Tree Paths/solution.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def binaryTreePaths(self, root):
        """
        :type root: TreeNode
        :rtype: List[str]
        """
        def dfs(root, path):
            if not root:
                return
            if not root.left and not root.right:
                result.append(path + str(root.val))
            path = path + str(root.val) + "->"
            dfs(root.left, path)
            dfs(root.right, path)
        result = []
        dfs(root, "")
        return result
